lib/github_service/commands/run_test.rb

```diff
@@ -0,0 +1,302 @@
+        Settings.github.run_tests_repo.name
+      end
+
+      def self.test_repo_clone_dir
+        @test_repo_clone_dir ||= begin
+                                   url_parts = test_repo_url.split("/")[-2, 2]
+                                   repo_org  = url_parts.first
+                                   repo_dir  = test_repo_name
+                                   File.join(::Repo::BASE_PATH, repo_org, repo_dir)
+                                 end
+      end
+
+      # Name of the bot
+      def self.bot_name
+        Settings.github_credentials.username
+      end
+
+      # The new branch name for this particular run of the command (uniq)
+      def branch_name
+        @branch_name ||= "#{self.class.bot_name}-run-tests-#{issue.number}-#{SecureRandom.uuid}"
+      end
+
+      def run_tests
+        clone_repo_if_necessary
+        create_cross_repo_test_branch
+        update_travis_yaml_content
+        commit_travis_yaml_changes
+        push_commit_to_remote
+        create_cross_repo_test_pull_request
+      end
+
+      private
+
+      def _execute(issuer:, value:)
+        @issuer = issuer
+        parse_value(value)
+        return unless valid?
+
+        run_tests
+      end
+
+      def parse_value(value)
+        @value = value
+
+        @test_repos, @repos = value.split(/\s+including\s+/)
+                                   .map { |repo_list| repo_list.split(",") }
+
+        # Add the identifier for the PR for this comment to @repos here
+        @repos ||= []
+        @repos  << "#{issue.repo_name}##{issue.number}"
+
+        @test_repos.map! { |repo_name| normalize_repo_name(repo_name) }
+        @repos.map!      { |repo_name| normalize_repo_name(repo_name) }
+      end
+
+      def normalize_repo_name(repo)
+        repo.include?("/") ? repo : "#{issue.organization_name}/#{repo}"
+      end
+
+      def valid?
+        unless issue.pull_request?
+          issue.add_comment("@#{issuer} 'run-test(s)' command is only valid on pull requests, ignoring...")
+          return false
+        end
+
+        unless repos_valid?(@test_repos + @repos)
+          issue.add_comment("@#{issuer} '#{value}' is an invalid repo, ignoring...")
+
+          return false
+        end
+
+        true
+      end
+
+      REPO_ONLY_REGEXP = /(?:[^\/]+\/)?(?<REPO_NAME_ONLY>[^@#]+)/.freeze
+      def repos_valid?(repos)
+        repos.all? do |repo|
+          name_only = repo.match(REPO_ONLY_REGEXP)[:REPO_NAME_ONLY]
+          GithubService.org_repos(issue.organization_name).include?(name_only)
+        end
+      end
+
+      ##### run_tests steps #####
+
+      # Clone repo (if needed) and initialize @rugged_repo
+      def clone_repo_if_necessary
+        repo_path = self.class.test_repo_clone_dir
+        if Dir.exist?(self.class.test_repo_clone_dir)
+          @rugged_repo = Rugged::Repository.new(repo_path)
+        else
+          url = self.class.test_repo_url
+          @rugged_repo = Rugged::Repository.clone_at(url, repo_path)
```

Comment:
While I like using clone_at (and we should definitely use this to replace `MiniGitService.clone`!), I think this may not be thread-safe.  That is, if 2 threads try to clone at the same time, the exists check is not atomic....I'm not sure Rugged is thread safe at clone time.

Reply:
Probably not, though, I specifically now remember that we were solving something like this for the `EmbeddedAnisble` work earlier this year, so I probably can just look at that for reference on what to do here 😄 .

Reply:
*THAT'S* where I remember it from...I knew we were hitting this elsewhere and just couldn't remember :)

Reply:
@Fryguy and myself had a chat about this PR (and others), and I think on this issue, I suggested that we don't both with thread safety for this currently since it for our use case, we should only be cloning once.  After that, it will just be a `.fetch`, which should be thread safe.

Moving forward, when we refactor a decent amount of the git logic in this repo, I think making a clone method that is thread/process safe is a good idea, just not something that is needed for a single repo usage like we have here.